### PR TITLE
refine paxos implementation and readability

### PR DESCRIPTION
**Pull request by changchengx**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/mon/Paxos.cc [hunk 1]
```diff
@@ -503,27 +503,25 @@ void Paxos::handle_last(MonOpRequestRef op)
...
-  for (auto p = peer_last_committed.begin();
-       p != peer_last_committed.end();
-       ++p) {
-    if (p->second + 1 < first_committed && first_committed > 1) {
+  for (const auto& [peon_rank, peon_last_committed] : peer_last_committed) {
+    if (peon_last_committed + 1 < first_committed && first_committed > 1) {
       dout(5) << __func__
-	      << " peon " << p->first
-	      << " last_committed (" << p->second
+	      << " peon " << peer_rank
```
Comment (tchaikov) [inline src/mon/Paxos.cc:509]: s/peer/peon/

File: src/mon/Paxos.cc [hunk 2]
```diff
@@ -503,27 +503,25 @@ void Paxos::handle_last(MonOpRequestRef op)
   ceph_assert(g_conf()->paxos_kill_at != 2);
 
   // is everyone contiguous and up to date?
-  for (auto p = peer_last_committed.begin();
-       p != peer_last_committed.end();
-       ++p) {
-    if (p->second + 1 < first_committed && first_committed > 1) {
+  for (const auto& [peon_rank, peon_last_committed] : peer_last_committed) {
```
Comment (tchaikov) [inline src/mon/Paxos.cc:506]: s/peon/peer/

Comment (changchengx) [reply src/mon/Paxos.cc:506]: After peon sends MMonPaxos::OP_LAST message to the leader, the leader deals with the message by Paxos::handle_last.
Previously, I also want to "s/peon/peer". However, the auto bind name "peon_last_committed" will be the same as "peer_last_committed" in that way.

File: src/mon/Paxos.cc (hunk 1 above)
Comment (tchaikov) [inline src/mon/Paxos.cc:509]: s/peer/peon/

File: src/mon/Paxos.cc [hunk 3]
```diff
@@ -503,27 +503,25 @@ void Paxos::handle_last(MonOpRequestRef op)
...
 	      << ") -- bootstrap!" << dendl;
       op->mark_paxos_event("need to bootstrap");
       mon->bootstrap();
       return;
     }
-    if (p->second < last_committed) {
+    if (peon_last_committed < last_committed) {
       // share committed values
-      dout(10) << " sending commit to mon." << p->first << dendl;
+      dout(10) << " sending commit to mon." << peer_rank << dendl;
```
Comment (tchaikov) [inline src/mon/Paxos.cc:519]: ditto.

File: src/mon/Paxos.cc [hunk 4]
```diff
@@ -503,27 +503,25 @@ void Paxos::handle_last(MonOpRequestRef op)
...
       MMonPaxos *commit = new MMonPaxos(mon->get_epoch(),
 					MMonPaxos::OP_COMMIT,
 					ceph_clock_now());
-      share_state(commit, peer_first_committed[p->first], p->second);
-      mon->send_mon_message(commit, p->first);
+      share_state(commit, peer_first_committed[peer_rank], peon_last_committed);
```
Comment (tchaikov) [inline src/mon/Paxos.cc:523]: and here.

Comment (changchengx) [reply src/mon/Paxos.cc]: @tchaikov 
Sorry for these build errors. I haven't built them locally. 
Currently, I'm studying the Paxos protocol in Ceph to add one new PaxosService in Monitor.

**Comment by changchengx** [comment]
@tchaikov Please help review this PR.
1. rename the variable to improve readability.
2. use range-loop to improve readability.
3. fix one little problem
4. add some comments about the difference between Ceph's Paxos and normal Paxos.
5. use std::move to avoid adding the reference number in the bufferlist.

**Comment by changchengx** [comment]
@tchaikov ping

**Comment by tchaikov** [comment]
@changchengx i am still on vacation. will take a look after next week.

**Comment by changchengx** [comment]
@tchaikov 
Enjoy your holiday time. Thanks for putting this into your pipeline.

**Comment by changchengx** [comment]
@tchaikov ping

**Comment by changchengx** [comment]
Please close this request.
Other Intel employees may work on it if this is still needed.

**Pull request closed without merging**